cobald_hep_plugins/timer.py

```diff
@@ -0,0 +1,92 @@
+"""Time-based demand control decorators and helpers."""
+
+from __future__ import annotations
+
+import asyncio
+from datetime import datetime, time as datetime_time
+from typing import Mapping, Union, Iterable
+
+from cobald.daemon import service
+from cobald.interfaces import Pool, PoolDecorator
+
+TimeInput = Union[str, datetime_time]
+Schedule = Mapping[TimeInput, float]
+
+
+def str_to_time(value: str) -> datetime_time:
+    """Convert a HH:MM string into a :class:`datetime.time` object."""
+    try:
+        hours_str, minutes_str = value.strip().split(":", 1)
+        hours, minutes = int(hours_str), int(minutes_str)
+    except ValueError as exc:
+        raise ValueError(f"invalid time specification {value}")
+    return datetime_time(hour=hours, minute=minutes)
+
+
+def latest_timestamp(
+    times: Iterable[datetime_time],
+    *,
+    reference: datetime_time | None = None, # easier to test with
+) -> datetime_time:
+    """Return the latest timestamp with respect to the current time."""
+    ordered_times = sorted(times) # ascending
+    current_time = datetime.now().time() if reference is None else reference
+    for candidate in reversed(ordered_times): # descending
+        if candidate <= current_time:
+            return candidate
+    # All configured times are in the future relative to ``current_time``.
+    # return latest timestamp
+    return ordered_times[-1]
+
+
+@service(flavour=asyncio)
+class Timer(PoolDecorator):
+    """
+    Decorator that adjusts demand based on a daily schedule.
+
+    :param target: pool being decorated
+    :param schedule: mapping of ``HH:MM`` times
+                     to the demand that should become active at that time
+    :param interval: interval in seconds between schedule checks
+    """
+
+    def __init__(
+        self,
+        target: Pool,
+        schedule: Schedule,
+        interval: int = 300,
+    ):
+        super().__init__(target)
+
+        assert interval > 0, "Interval must be a positive integer."
+        self.interval = interval
+
+        schedule = {str_to_time(key): value for key,value in schedule.items()}
+        self.schedule = schedule
```

Comment:
In `_refresh_demand` is an assertion that the `value` here is positive. Please do the validation here already. Since the data must be sorted for various algorithms, either checking or enforcing this here seems prudent.  
I also recommend to rename the `key`/`value` to what they actually represent.

```suggestion
        schedule = {str_to_time(time): demand for time, demand in schedule.items()}
        assert all(demand >= 0 for demand in schedule.values())
        self.schedule = dict(sorted(schedule.items()))
```